google/resumable_media/_upload.py

```diff
@@ -860,6 +869,375 @@ def recover(self, transport):
         raise NotImplementedError("This implementation is virtual.")
 
 
+class XMLMPUContainer(UploadBase):
+    """Initiate and close an upload using the XML MPU API.
+
+    An XML MPU sends an initial request and then receives an upload ID.
+    Using the upload ID, the upload is then done in numbered parts and the
+    parts can be uploaded concurrently.
+
+    In order to avoid concurrency issues with this container object, the
+    uploading of individual parts is handled separately, by XMLMPUPart objects
+    spawned from this container class. The XMLMPUPart objects are not
+    necessarily in the same process as the container, so they do not update the
+    container automatically.
+
+    MPUs are sometimes referred to as "Multipart Uploads", which is ambiguous
+    given the JSON multipart upload, so the abbreviation "MPU" will be used
+    throughout.
+
+    See: https://cloud.google.com/storage/docs/multipart-uploads
+
+    Args:
+        upload_url (str): The URL of the object (without query parameters). The
+            initiate, PUT, and finalization requests will all use this URL, with
+            varying query parameters.
+        headers (Optional[Mapping[str, str]]): Extra headers that should
+            be sent with the :meth:`initiate` request, e.g. headers for
+            encrypted data. These headers will be propagated to individual
+            XMLMPUPart objects spawned from this container as well.
+
+    Attributes:
+        upload_url (str): The URL where the content will be uploaded.
+    """
+
+    def __init__(self, upload_url, filename, headers=None, upload_id=None, parts=None):
+        super().__init__(upload_url, headers=headers)
+        self._filename = filename
+        self._content_type = None
+        self._upload_id = upload_id
+        self._parts = parts if parts is not None else {}
+
+    @property
+    def upload_id(self):
+        return self._upload_id
+
+    def register_part(self, part_number, etag):
```

Comment:
They need to be independent because they run in separate processes - the container is in the main process and the part is in a subprocess. Integrating them will be handled by python-storage transparently. This is just an API that facilitates that. I don't expect that anyone will want to use it separately from python-storage, but if they choose to do so, the docstrings will help them.